File: PHP/One/Transforms/UseDefaultCallArgsExplicitly.php
<?php

namespace One\Transforms\UseDefaultCallArgsExplicitly;

use One\Ast\Expressions\Expression;
use One\Ast\Expressions\IMethodCallExpression;
use One\Ast\Expressions\InstanceMethodCallExpression;
use One\Ast\Expressions\NewExpression;
use One\Ast\Expressions\StaticMethodCallExpression;
use One\Ast\Types\IMethodBase;
use One\Ast\Types\IMethodBaseWithTrivia;
use One\Ast\Types\MethodParameter;
use One\AstTransformer\AstTransformer;

class UseDefaultCallArgsExplicitly extends AstTransformer {
    function __construct() {
        parent::__construct("UseDefaultCallArgsExplicitly");
        
    }
    
    protected function getNewArgs($args, $method) {
        if (array_key_exists("UseDefaultCallArgsExplicitly", $method->attributes) && @$method->attributes["UseDefaultCallArgsExplicitly"] ?? null === "disable")
            return $args;
        if (count($args) >= count($method->parameters))
            return $args;
        
        $newArgs = array();
        for ($i = 0; $i < count($method->parameters); $i++) {
            $init = $method->parameters[$i]->initializer;
            if ($i >= count($args) && $init === null) {
                $this->errorMan->throw("Missing default value for parameter #" . $i + 1 . "!");
                break;
            }
            $newArgs[] = $i < count($args) ? $args[$i] : $init;
        }
        return $newArgs;
    }
    
    protected function visitExpression($expr) {
        parent::visitExpression($expr);
        if ($expr instanceof NewExpression && $expr->cls->decl->constructor_ !== null)
            $expr->args = $this->getNewArgs($expr->args, $expr->cls->decl->constructor_);
        else if ($expr instanceof InstanceMethodCallExpression)
            $expr->args = $this->getNewArgs($expr->args, $expr->method);
        else if ($expr instanceof StaticMethodCallExpression)
            $expr->args = $this->getNewArgs($expr->args, $expr->method);
        return $expr;
    }
}

File: CSharp/One/Transforms/UseDefaultCallArgsExplicitly.cs
using One.Ast;
using One;
using System.Collections.Generic;

namespace One.Transforms
{
    public class UseDefaultCallArgsExplicitly : AstTransformer {
        public UseDefaultCallArgsExplicitly(): base("UseDefaultCallArgsExplicitly")
        {
            
        }
        
        protected Expression[] getNewArgs(Expression[] args, IMethodBaseWithTrivia method)
        {
            if (method.attributes.hasKey("UseDefaultCallArgsExplicitly") && method.attributes.get("UseDefaultCallArgsExplicitly") == "disable")
                return args;
            if (args.length() >= method.parameters.length())
                return args;
            
            var newArgs = new List<Expression>();
            for (int i = 0; i < method.parameters.length(); i++) {
                var init = method.parameters.get(i).initializer;
                if (i >= args.length() && init == null) {
                    this.errorMan.throw_($"Missing default value for parameter #{i + 1}!");
                    break;
                }
                newArgs.push(i < args.length() ? args.get(i) : init);
            }
            return newArgs.ToArray();
        }
        
        protected override Expression visitExpression(Expression expr)
        {
            base.visitExpression(expr);
            if (expr is NewExpression newExpr && newExpr.cls.decl.constructor_ != null)
                newExpr.args = this.getNewArgs(newExpr.args, newExpr.cls.decl.constructor_);
            else if (expr is InstanceMethodCallExpression instMethCallExpr)
                instMethCallExpr.args = this.getNewArgs(instMethCallExpr.args, instMethCallExpr.method);
            else if (expr is StaticMethodCallExpression statMethCallExpr)
                statMethCallExpr.args = this.getNewArgs(statMethCallExpr.args, statMethCallExpr.method);
            return expr;
        }
    }
}